Dropout.lua

```diff
@@ -20,13 +21,14 @@ function Dropout:updateOutput(input)
       self.output:resizeAs(input):copy(input)
    end
    if self.p > 0 then
-      if self.train then
-         self.noise:resizeAs(input)
-         self.noise:bernoulli(1-self.p)
+      local si = self.si and not self.train
+      if self.train or si and not self.v2 then
+         local noise = si and input:clone() or self.noise:resizeAs(input)
```

Comment:
If you want to save memory during inference, it is usually better to just share the `noise` field among other dropout layers or other internal buffers from other modules.
Currently, this line is actually allocating new memory at every iteration, and freeing it just afterwards, which is a blocking operation on GPU. We usually try to avoid that as much as possible in `nn`.